### ISSUE TITLE
Nodes removed from the cluster may prevent features from being enabled for A_VERY_LONG_TIME (3 days)

### ISSUE THREAD

**Issue by piodul** [scylladb/scylladb#14194]
HEAD: d2e089777bf5cca87a81a6ace7c5d67d37365a08

Currently, a node will attempt to enable features when it notices that the `SUPPORTED_FEATURES` state of some of the nodes have been changed. A set of features that is common to all nodes is calculated based on the system.peers state and the gossip state:

https://github.com/scylladb/scylladb/blob/d2e089777bf5cca87a81a6ace7c5d67d37365a08/gms/gossiper.cc#L2402-L2453

The problem is that, as seen above, we are considering features of all nodes that are in the `_endpoint_state_map` and that includes nodes that were removed from the cluster. The gossiper keeps removed nodes in that map for three days (as defined by the `A_VERY_LONG_TIME` constant) before erasing this information from the map. Effectively, if the cluster is upgraded within three days, there is a chance that features won't be enabled. After this period ends, the new features won't be enabled until a notification about changing `application_state::SUPPORTED_FEATURES` is triggered (which can happen e.g. due to node restart/joining a new node).

I think we should consider not including nodes that have left the cluster when calculating features that should be enabled.

cc: @asias

**Comment by kostja** [comment]
Only in centralized topology changes. This has a reason to be done this way in gosisp.

**Comment by kbr-scylla** [comment]
> I think we should consider not including nodes that have left the cluster when calculating features that should be enabled.

Do we even include them in raft-topo mode? IIRC we base features on NORMAL nodes.

**Comment by piodul** [comment]
> > I think we should consider not including nodes that have left the cluster when calculating features that should be enabled.
> 
> Do we even include them in raft-topo mode? IIRC we base features on NORMAL nodes.

No, we don't do it in raft mode. This issue is only relevant to the gossip mode.

**Comment by kbr-scylla** [comment]
blocked by https://github.com/scylladb/scylladb/issues/15422

**Comment by mykaul** [comment]
> blocked by [#15422](https://github.com/scylladb/scylladb/issues/15422)

The above is merged.

**Comment by piodul** [comment]
> > blocked by [#15422](https://github.com/scylladb/scylladb/issues/15422)
> 
> The above is merged.

Yes, the issue is no longer relevant (starting from 2026.2). I'm closing the issue.
